### ISSUE TITLE
OpenId JSON::JWS::VerificationFailed

### ISSUE THREAD

**Issue by SavannahDearing** [onc-healthit/inferno#377]
**Subject of the issue**
Getting JSON::JWS::VerificationFailed on testing OpenId ID Token when I can verify signature with own code and using third party jwt.io

![openidError](https://user-images.githubusercontent.com/58518170/70150325-d9d8ba00-1677-11ea-9475-d90a6d11954c.png)

**Your environment**
* Edition of inferno (Community or Program): Web
* Version of inferno: Proposed ONC Health IT Certification FHIR DSTU2
* Which browser and version(s) is the bug present on?: Chrome

**Steps to reproduce**
IdToken:
ewoJImFsZyIgOiAiUlMyNTYiLAoJInR5cCIgOiAiSldUIgp9.ewoJImF1ZCIgOiAiZmU3YzgzMDM3NTlkMWQiLAoJImV4cCIgOiAiMTU3NTM5MjA0OCIsCgkiZmhpclVzZXIiIDogImh0dHBzOi8vd3d3Lm1lZGVudG1vYmlsZXRlc3QuY29tL2ZoaXIvRFNUVTIvc2F2Y3cyMzUvUGF0aWVudC8xMjM0IiwKCSJpYXQiIDogIjE1NzUzOTE3NDgiLAoJImlzcyIgOiAiaHR0cHM6Ly93d3cubWVkZW50bW9iaWxldGVzdC5jb20vZmhpci9EU1RVMi9zYXZjdzIzNSIsCgkic3ViIiA6ICJDRDgwMDVGNTc0MzI2M0VDQjBFQzE1NzUzOTE3NDgiCn0.EAazBFLIeW-0DB-DSsoFYp099QlVoFIYg8XMhFpY2pdESUjvW-IvV1BX0BZhrGxgMRgfkdmJ6ehsqwGBsFWhU28vgbk47Lg5cgIAckglQseDt3eQbqrXy6nCYTQottQdaOfZZCrsw9c57_0XfztdBXADN-Pxkw9c8pMalMUXCrg2NrO7XffUCQNpc1TG2O1yfDp9-2lfE0PIjS5hf9zrxEY7FBph4zavOnmEBAqmvOEqMRnroPbOdZBK8KTSETP57cuYq6QBzYLAyH_OBYALNrlcq9VDIEsoqnoBuAiD0K40R27dxoGY0oWykncZdOZYJHCOtmdhcGKmG-aP3O6AccIDgxD6tDuE_midfKBY6fUR7ZMfkF6k9ndaWcrUky3wIcSM73Fpzf1ffnV7lxDIrNDh3-5YN5DHHvCfrCDGKHHeaSC3VwJV6zeA1O8Wr6wOWnxaStM4pSw6yyrz8lzlpZmfPvypFb3UmaVXiGi092Pty8spnaFHrENwQo2ubNLe

Public Key:
-----BEGIN RSA PUBLIC KEY-----
MIIBiAKCAYEA8fCdlJwwoPRoT1+eyBYgG1BQEq4sQJkiioBQ4lMyTlu9ToX1vkjRqFvhTbcQ8E+mlaVofub0ZxOXp//y0yMASEs80CWNuoAjz9W//MhwZGErp0lyMSZnddeDrNO6rMQl9R60i+M4jz4FKIspzznMiM30vM+le4GgIENckguhiGpAl35NwRfL/Q+py0gmUb0NxpHPa+wLjJ0iraB4j0aaKVV7QU8IzGsia3CApiwVPd7eAdNr1ex+/pum7t3mf9HeXHSNJkQ6QvHOJAb4VpReZfRIxc4zvozJn9YkCASh8+C7vJ8G2bSeqcVxIkN+CJLn3EHD0rWplrQCVaYMFQ9MiAa4HlFBBMN2Ka6xctXK1hSDGJhpUYYJJOBBCw1r9ssXe5k+VwWcX0rYp3wo6wJe/6SFnqn7SxhOVjkw9psxMSTWKVbqTk8lFD7XZKlxb0BLEjOmWS1UPaTOtpucznrGdp2Z6D0aWxak6BV7E2uEGqPP6B0BjBpM8qjVJF8zE4sXAgER
-----END RSA PUBLIC KEY-----

JWK URI:
https://www.medentmobiletest.com/fhir/DSTU2/jwks/index.php?medent_practice_id=savcw235

If I use https://jwt.io/ and plug in the IdToken and Public Key, and select RS2526 as the Algorithm, the signature is verified.

![openidSigVerified](https://user-images.githubusercontent.com/58518170/70150526-23290980-1678-11ea-8f35-dad0bbe9572f.png)

This is the first time I've tried to implement OpenId so I'm wondering if maybe I'm missing a parameter in the JWK or ID Token that's preventing the signature from validating, or maybe Inferno doesn't support RSA256?  The error message is not very clear to me on why it does not validate.

**Comment by Jammjammjamm** [comment]
It does look like RSA256 support might be the issue. It works when I convert the public key to the following jwk:
```
{
  "kty": "RSA",
  "e": "EQ",
  "n": "8fCdlJwwoPRoT1-eyBYgG1BQEq4sQJkiioBQ4lMyTlu9ToX1vkjRqFvhTbcQ8E-mlaVofub0ZxOXp__y0yMASEs80CWNuoAjz9W__MhwZGErp0lyMSZnddeDrNO6rMQl9R60i-M4jz4FKIspzznMiM30vM-le4GgIENckguhiGpAl35NwRfL_Q-py0gmUb0NxpHPa-wLjJ0iraB4j0aaKVV7QU8IzGsia3CApiwVPd7eAdNr1ex-_pum7t3mf9HeXHSNJkQ6QvHOJAb4VpReZfRIxc4zvozJn9YkCASh8-C7vJ8G2bSeqcVxIkN-CJLn3EHD0rWplrQCVaYMFQ9MiAa4HlFBBMN2Ka6xctXK1hSDGJhpUYYJJOBBCw1r9ssXe5k-VwWcX0rYp3wo6wJe_6SFnqn7SxhOVjkw9psxMSTWKVbqTk8lFD7XZKlxb0BLEjOmWS1UPaTOtpucznrGdp2Z6D0aWxak6BV7E2uEGqPP6B0BjBpM8qjVJF8zE4sX",
  "kid": "AI2FrnQNyQltFMiNQQiz3jkcphn1R-iS80wORcXjUmk"
}
```
It doesn't work with the jwk from the url you provided, though, and I get the same error message that you posted. We'll look into this issue, but it might take some time.

**Comment by SavannahDearing** [comment]
If it helps, I'm using Cryptopp to sign the message:
```
std::string SignMessageWithRsaKeyAndB64Encrypt(CryptoPP::RSA::PrivateKey privateKey, std::string sSignMessage){
   CryptoPP::RSASS<CryptoPP::PKCS1v15, CryptoPP::SHA256>::Signer signer(privateKey);

   CryptoPP::AutoSeededRandomPool rng;
   std::string sSignature;
   CryptoPP::StringSource ss(sSignMessage, true,
                               new CryptoPP::SignerFilter(rng, signer,
                                   new CryptoPP::StringSink(sSignature)));

   sSignature = cnvbase64::EncodeB64URLwCrypto(sSignature);

   return sSignature;
}
```

Then to Self-Validate:
```
void CheckJWK(std::string n, std::string e){

   ////////////////////////////////////////////////////////////////////
   //converting n and e from JWK to PEM to validate in jwt.io:
   std::string decryptN = cnvbase64::DecodeB64URLwCrypto(n);
   std::string decryptE = cnvbase64::DecodeB64URLwCrypto(e);
   CryptoPP::Integer nn(decryptN.c_str(), CryptoPP::BIG_ENDIAN_ORDER);
   CryptoPP::Integer ee(decryptE.c_str(), CryptoPP::BIG_ENDIAN_ORDER);
   CryptoPP::RSA::PublicKey pubKey;
   pubKey.Initialize(nn, ee);

   std::string stringkey;
   pubKey.DEREncodePublicKey(CryptoPP::StringSink(stringkey).Ref());
   AddTroubleshootingLog("DER key for jwt.io check: " + cnvbase64::EncodeB64wCrypto(stringkey));

   ////////////////////////////////////////////////////////////////////
   //self-validate with cryptopp:

   //generate idToken for testing:
   std::string idToken = FhirAppAuthorization::GenerateIdTokenForOpenId("clientid", "patientmrn");
   
   //unpack idToken as JWT, This seperates and base64URL decodes each piece
   std::string sSignatureMessage, sHeader, sPayload, sSignature;
   FhirAuthorizationRSAKeysAndJWT::UnpackJwt(idToken, sSignatureMessage, sHeader, sPayload, sSignature );

   //loading keys from storage:
   CryptoPP::RSA::PublicKey publicKey;
   LoadPatientAccessOpenIdRSAKey(publicKey);

   //verify signature with cryptoPP:

   CryptoPP::RSASS<CryptoPP::PKCS1v15, CryptoPP::SHA256>::Verifier verifier(publicKey);

   bool bSignatureVerified = false;

   CryptoPP::StringSource ss2(sSignature + sSignatureMessage, true,
                              new CryptoPP::SignatureVerificationFilter(
                                verifier,
                                new CryptoPP::ArraySink((CryptoPP::byte*)&bSignatureVerified,
                                sizeof(bSignatureVerified)
                                )
                              )
   );

   if(!bSignatureVerified)
      AddTroubleshootingLog("signature not valid");
   else
      AddTroubleshootingLog("signature OK");
}
```

I believe PKCS1v15 / SHA256 is what we're supposed to be using but the documentation I've seen on this seems to be a bit elusive on these details.  We've at least been able to validate signatures based on the jwt.io format [choosing Algorithim RS256]  using this method.

Thanks for your prompt response :)

**Comment by Jammjammjamm** [comment]
This has hopefully been resolved by #391, which has not yet been deployed. This update reworked a lot of the OIDC sequence to more closely match the spec, so it's possible you will encounter other new failures.

**Comment by Jammjammjamm** [comment]
Marking as closed per: https://github.com/onc-healthit/inferno-program/issues/80
